### PR TITLE
[wip] GitLab evaluation

### PR DESCRIPTION
**Pull request by aparcar**
This does some basic formal testing, just like done for packages.git.
Later compiling of device tree files, checking kmods or other more
specific tests could be done. However this is a first step to CI.

WIP: not fully tested yet

Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

**Comment by ynezz** [comment]
Thanks a lot for your persistence, patience and your continuous effort on the CI improvements!

Anyway, as we're probably going to [consider switch to GitLab](https://openwrt.org/meetings/hamburg2019/start#defragmentation_of_code_and_tools) I would really like to avoid any tighter integration with this proprietary services and would rather focus our work towards GitLab integration instead. It's also going to be much easier for you, as you already have all necessary admin rights there, right?

I envision following TODO list:

- [ ] for the start, create just unidirectional GitHub -> GitLab mirroring bridge and start mirroring PRs from GitHub to our testing GitLab instance
- [ ] setup CI and other tools properly at our GitLab instance so we could copy&paste the CI check results back here to the GitHub until we iron out the corner cases
- [ ] once the CI checks are good enough, we could enable bidirectional GitHub <-> GitLab bridge and propagate back GitLab CI check results to GitHub as PR comments (or via some other possible integration hooks)
- [ ] integrate Patchwork/mail workflow with the GitLab instance in the same spirit
- [ ] cast a vote for making our GitLab instance our main platform for development
- [ ] migrate bugs from FlySpray to GitLab issues (this is probably going to be little bit painful as FlySpray has no API access), redirect bugs.openwrt.org to gitlab.openwrt.org/openwrt/issues (or similar), shutdown bugs.openwrt.org

**Comment by aparcar** [comment]
Thanks for the writeup, I'll meet lynxis today and try to get some stuff done

**Comment by diizzyy** [comment]
At least do a public query of what people in general want if you want to attract people instead of bulldozing over everyone like last time. Why is there a need to change at all?

**Comment by ynezz** [comment]
tl;dr Don't worry, there's no plan to leave GitHub.

> At least do a public query of what people in general want if you want to attract people

Why would we need to do a public query? This is more about team members, core developers. We're just seeing a lot of issues here and there and we would like to fix them or at least attempt to. Some of us simply believe, that GitLab is probably going to help with this regard. You can see the list of the main pain points at the [Hamburg 2019 meeting summary](https://openwrt.org/meetings/hamburg2019/start).

> instead of bulldozing over everyone like last time.

By last time, you mean, when the project moved from SourceForge to GitHub? :-) Putting jokes aside, we first want to play with the GitLab little bit and once we think, that it would be good to proceed, then we're going to vote about it. It's a long term goal, I don't expect, that it would happen this year. We're probably going to make a summary and another talk about it first during next meeting in 2020.

> Why is there a need to change at all?

* unify and integrate currently fragmented workspace (patchwork, GitHub, Flyspray) at one place
* faster/customizable workflow
* boost QA/CI integration, automate patch/PR review processes
* promote, use & develop other OSS project

Someone has asked about [this topic on IRC](http://logs.nslu2-linux.org/livelogs/openwrt-devel/openwrt-devel.20190716.txt) so you can find more details there as well, it starts at `Jul 15 09:20:31`.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      - BRANCH: "master"
+    steps:
+      - checkout:
+          path: ~/openwrt_packages
+
+      - run:
+          name: Check changes / verify commits
+          working_directory: ~/openwrt_packages
+          command: |
+             cat >> $BASH_ENV <<EOF
```
Comment (ynezz) [inline .circleci/config.yml]: I think, that we've discussed this already, nobody sane is going to suffer with YAML pox voluntarily :-) so it would be better to keep the YAML file as short as possible and do the rest in the external Bash/Python/Perl scripts instead (probably in separate directory `scripts/ci`) ? Ideally those checks could be run as make targets `make ci-formal-checks` (defined in `include/ci-commands.mk`) even localy by the developers, so they could simply check the contributions before sending patch/PR, debug/develop those CI checks etc. This way it's probably also CI agnostic, so in the end it doesn't matter if it's run on CircleCI or GitLab CI or whatever we choose in the future.

Comment (aparcar) [reply .circleci/config.yml]: Good point, I'll surely work on that

**Comment by diizzyy** [comment]
@ynezz 
One of the many reason for the split and merge was to become more open and attract more developers/contributors. This appears to have regressed quite a bit in the last months and things are starting to go into a more closed group thing once again but I guess this boils down to if you find yourself more valuable than the project itself. It's also very tiresome when some unwritten rule pops up out of nowhere and everything comes to a grinding halt.

Doing a survey like FreeBSD did would be a better idea rather than "we think" at least if you value the community and contributors.. 
https://youtu.be/9nc8N6GtAPg?t=1101
https://youtu.be/9nc8N6GtAPg?t=1195 (Q15)
https://youtu.be/9nc8N6GtAPg?t=1306 (Q14)

**Comment by CodeFetch** [comment]
@diizzyy It's this developer meritocrazy thing... Give up your life, commit as much as possible no matter how useless the commits are and you are the robot king of FOSS development. You can become a steering gear, but after all you will need to become a gear.

**Comment by bobafetthotmail** [comment]
> @diizzyy It's this developer meritocrazy thing... Give up your life, commit as much as possible no matter how useless the commits are and you are the robot king of FOSS development. You can become a steering gear, but after all you will need to become a gear.

In what projects you could commit your way to leadership like that? Asking for a friend

**Comment by CodeFetch** [comment]
The only thing I wanted to say is that FOSS development is neither a question of democracy nor aristocracy and ends in a technocracy in the cloak of anarchy and meritocracy. Those who do, decide what get's done and those who decide can do more what they like to do and thus do more and get affirmation from their dev pals who act as they do and thus do more. There is a German word for this: "Glaubensgefängnis" - faith prison.

diizzyy's displeasure has a certain justification. If you don't ask all people involved, users and devs about or at least try to consider what they want, because they are not the "main devs", the question is: For who do you do FOSS projects? In that case only for yourselves... You are free to do this, but don't think you are selfless, when you do it. And don't expect people to be happy with your decisions. Similarly politicians live in a faith prison - why all of our governments kind of suck. The solution is simple. Don't reply to displeasure with "We do it, because we do it" in a shady long text. That's what politicians do and we should be better than them, because we have computers and free access to knowledge. 

ynezz, nothing against you. You are a great person. Just trying to show up a communication problem here...

> But as I'm personally very strongly against **b)** I'm not going to prepare such survey (or even think about it for a moment)

Sounds like ynezz doesn't, so it's likely never going to happen. Actually many people love using Github (and ynezz even pays them, because of laziness) and people more likely use Github issues than the bugtracker, more likely do pull requests than sending patches via email and more likely fork this project, all because of laziness. In this case laziness leads to survival of the fittest and that is Github.

> "Those Who Do Not Learn History Are Doomed To Repeat It." (BitKeeper, SourceForge, Google Code [to name a few examples](https://mako.cc/writing/hill-free_tools.html)) :-)

All very complicated solutions compared to Github. Nothing for lazy people. That's why they starved. Github is better (at least its UI) and thus does not repeat history.

Don't understand me wrong. Personally I prefer Gitlab. Personally I don't like iPhones, Windows, Facebook, Instagram, Twitter and so on... But we need to take the people who use and like it serious and try to understand them. If we think we are more qualified than these people to decide that their solutions are wrong or their opinion does not count, we will loose them. And in the end we live in a Glaubensgefängnis. That's why Linux is still not mainstream. That's why OpenWrt is still not mainstream. I remember an interview with Karl R. Popper where he said something like: "The problem with any ideology is that those who push it forward hope for a greater impact.".

**Comment by bobafetthotmail** [comment]
> If you don't ask all people involved, users and devs about or at least try to consider what they want, because they are not the "main devs", the question is: For who do you do FOSS projects?

That's the crucial point here, people keep thinking that FOSS projects have some kind of obligation to listen to people, or that they are somehow more "democratic" or "selfless" or something.

No they don't. FOSS only means you can fork the code and go on your merry way without issues, and contribute back if your view aligns with the project developers.

It is and will remain some form of oligarchy where the core developers have most power, and contributors have less, random users have near zero as usual.

The goal of most FOSS projects is just to pool development resources between like-minded individuals/organizations, not to be a political system that chooses "what is best for its users".

LEDE at the time of the fork did say that being more transparent was part of their goals, and theoretically this should also still hold true in the "new OpenWrt", but that's their own policy, not a FOSS requirement per-se.

> Sounds like ynezz doesn't, so it's likely never going to happen. 

Another important aspect people commonly get wrong in FOSS. None is forced to do anything, so if you really want something it's YOU that have to go and talk to people and do stuff.

ynezz said he does not have enough motivation to actually change the status quo. That's fine. Are you aware of the time and effort required to do a proper survey?

If you are interested in these things, the best way is doing them yourself, not shame-guilting people to do them on your behalf.

> In this case laziness leads to survival of the fittest and that is Github.

And the last important point people commonly get wrong. FOSS projects are in most cases seriously undermanned and rely on contributor's spare time to do anything. SOME ones do have commercial entities and paid developers to pull the cart, but this isn't the case for OpenWrt.

"Laziness" is a very strong word in this case, even core developers have a very limited amount of time to devote to the project, I think a better term is "efficiency".

**Comment by CodeFetch** [comment]
Indeed... You've got the point of what I wanted to say. We just need to be aware of these things to not scare people away with taking this for granted and aim to be better (if we have the time to). Most developers were users at some point and they wouldn't have started working on those projects if they wouldn't like them. Thus a successful FOSS project needs to be user focused.

> * cast a vote for making our GitLab instance our main platform for development

This is a very important point and a good way to go.

> * migrate bugs from FlySpray to GitLab issues

And this should be reconsidered as mostly users file issues and a system preferably without authentication that is simple for users and not so much devs is needed.

But those are just my two cents.

**Comment by ynezz** [comment]
> > "Those Who Do Not Learn History Are Doomed To Repeat It." (BitKeeper, SourceForge, Google Code [to name a few examples](https://mako.cc/writing/hill-free_tools.html)) :-)
> 
> All very complicated solutions compared to Github. Nothing for lazy people. That's why they starved. Github is better (at least its UI) and thus does not repeat history.

Indeed, Microsoft is taking it to next level https://github.com/tkashkin/GameHub/issues/289

**Comment by filips123** [comment]
Do you have any specific reason why do you want to switch to GitLab? GitHub has a lot more users and features so it works be better to transfer all current bug trackers and pull requests to it.

If this is because US sanctions, also don't switch to GitLab. It is also US based and hosted on Google servers. And it's even worse because Google is one of the biggest corporations and they also have to obey US sanctions.

Real solution would be to switch to self-hosted GitLab, Gitea, Gogs or other open source program. For more ideas and details see [this Reddit topic](https://www.reddit.com/r/ethereum/comments/ciel0c/decentralised_alternative_to_github_built_on/).

**Comment by diizzyy** [comment]
X is going to be dropped/moved from tree or we're changing X to Y is just a few things and all have a tendency to be "announced" indirectly. Having little to no communication at all doesn't help and are all on the same page or is it something a few individuals have decided? 

I do think @CodeFetch's points are crucial for a healthy community in the long run but you are free to pick any direction you want.

As for consolidating services I think it's a great idea but everyone (especially main devs) or at least the vast majority needs be in agreement otherwise you'll have the same issue(s) as you have today. You can of course bulldoze everyone but that's no really ideal.

@ynezz 
If you want, I can ask but in short they asked (regarding contribution) how much time people spent, what CVS they used, what they prefered and if changing X to Y would increase their amount of contributions.

**Comment by claell** [comment]
Nice to see this is evaluated. I like their issue tracker and think that having everything in one place will improve the situation.

@diizzyy I am unsure, whether you understand the situation correctly. This is only for evaluation and there will be a voting, although I am not completely sure, this will be public:

> Putting jokes aside, we first want to play with the GitLab little bit and once we think, that it would be good to proceed, then we're going to vote about it.

**Comment by diizzyy** [comment]
@claell 
..and if you go one step further? Historically it's pretty much never been the case which exactly isn't keeping the community together.

**Comment by claell** [comment]
I don't know about the history, but there is no reason not to believe that behaviour would not change.

Additionally I want to mention that according to the chat protocol that @ynezz linked there are no plans to shutdown GitHub, but to keep the service running, like it is currently. So the only affected services I know are the issue tracker and the git hosting. Is this the things you want to keep unchanged or were you just fearing that they will shutdown GitHub?

**Comment by adrianschmutzler** [comment]
@aparcar @ynezz Since this was mentioned earlier, I would be interested in taking part ("be invited") in the GitLab evaluation (if a reasonable point for that is reached). I can't promise that I will be helpful, but I would be interested in following the process ...

**Comment by ynezz** [comment]
> @aparcar @ynezz Since this was mentioned earlier, I would be interested in taking part ("be invited") in the GitLab evaluation (if a reasonable point for that is reached). I can't promise that I will be helpful, but I would be interested in following the process ...

Invitation sent.

**Comment by aparcar** [comment]
I created a flyspray migration script parsing issues + comments from bugs.openwrt.org and adding them to the [evaluation GitLab](https://code.fe80.eu/openwrt/openwrt/issues), the [scraper](https://code.fe80.eu/openwrt/gitlab-evaluation/blob/master/flyspray/scraper.py) is using the [GitLab python lib](https://python-gitlab.readthedocs.io/) and [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/). Anybody who cares, the script output looks like [this](https://paste2.org/z3Z7Djf5)

**Comment by claell** [comment]
Nice! I thought GitLab had importing tools for most issue trackers ready, but apparently that is not the case for flyspray. I just had a quick look and noticed unwanted line breaks around urls.

Also there is apparently at least one other project I found that has migrated flyspray issues to their GitLab instance. It might be helpful to get in touch with them, in order to make use of some parts from their migration tools. 

PS: I'd also be interested in an invite.

**Comment by jow-** [comment]
> I created a flyspray migration script parsing issues + comments from bugs.openwrt.org

Nice work! I can give you SQL dumps in case it helps...

**Comment by aparcar** [comment]
SQL dumps would be great, parsing is the only hard part here. That should also resolve the issues mentioned by @claell

**Comment by lynxis** [comment]
Just to state it here as well. If you like to contribute to the gitlab evaluation. Please write me a mail lynxis@fe80.eu I'll send you an invitation back. We could also use a different gitlab instance (for testing!). I just offered my instance because I'm running it anyway. But it's only a small instance. The openwrt project at the gitlab instance is public. If you just want to look, you can.  https://code.fe80.eu/openwrt/

**Comment by CodeFetch** [comment]
What happens on e.g. a username conflict with OAuth? 

What I don't like about Gitlab/Github issues is that it's impossible to create issues without registration, because the average end-user won't register for doing that...

**Comment by jow-** [comment]
> What I don't like about Gitlab/Github issues is that it's impossible to create issues without registration

Our past experience with anonymous tickets was rather suboptimal. You often get drive-by style reports about issues which lack any relevant detail and no way to get back to the submitter because the issue was posted anonymously.

**Comment by CodeFetch** [comment]
@jow- That's a matter of implementation. Being able to file a bug report just by entering one's email and having to confirm the email for publication is less demotivating than registering, confirming one's email and only then being able to file a bug report. Next time forgetting one's password, requesting a password recovery email, setting a new password and then filing a bug report. 
Many users don't expect to ever file a bug report afterwards for that project again and are too lazy to register and don't have a Github account.

But it's out-of-scope for discussion as only flyspray supports this seemingly.

**Comment by claell** [comment]
> We could also use a different gitlab instance (for testing!).

You could open a test instance at gitlab.com. Or do you prefer to do hosting on your own servers (even for testing)? I'd also propose to think about moving completely to hosting at gitlab.com to save ressources after this evaluation period.

> What happens on e.g. a username conflict with OAuth?

I assume you mean when signing in with OAuth and your (old) username and password from flyspray or the OpenWrt website (not sure whether this functionality is even planned) and there is already someone with your username who signed up using a different method? I guess there is some mechanism preventing duplicate usernames like making you choose the username for the GitLab platform and making it independent from the login username.

> registering, confirming one's email and only then being able to file a bug report.

Luckily GitLab offers some variety of Sign-in options including GitHub, GitLab.com (if you are running on a self hosted instance) etc. While I prefer to create an account for each platform (and will do that also for creating bug reports) that other options will hopefully suit the people who just want to quickly file a bug. It might be harder to find out about that options and having to sign in with some authentication might push away some people, but regarding convenience that mechanism is probably even better than your proposal because there is no need to confirm the mail address.

**Comment by claell** [comment]
> Also there is apparently at least one other project I found that has migrated flyspray issues to their GitLab instance.

Link: https://gitlab.dune-project.org/flyspray/FS

Example of an issue: https://gitlab.dune-project.org/flyspray/FS/issues/442

What I like about it is that they apparently managed somehow to detect users on the GitLab instance and assigned their posts on Flyspray to them when available. Else they used the Flyspray Importer account. Also I like the style how they have the comments and description displayed. And they also migrated attachments.

**Comment by claell** [comment]
After some deeper digging I managed to find the relevant conversation which also contains their migration scripts: https://lists.dune-project.org/pipermail/dune-devel/2015-November/001659.html

**Comment by aparcar** [comment]
I created an evaluation instance [here](https://gitlab.com/openwrtorg/openwrt/issues) and currently creating a migration script from the bugs.openwrt.org sql to the gitlab issues. I'm trying to mirror all git.openwrt.org, however would be happy if somebody joins in. Please write me a mail for access.

**Comment by stangri** [comment]
@ynezz @aparcar since you gentlemen are investing some serious time in exploring this option, I'm wondering if you could address some of the questions:
1. Is the goal to explore the possibility of moving from proprietary SaaS GitHub to own instance of gitlab?
2. AFAIK even self-hosted gitlab instances require [monthly per-user fees](https://about.gitlab.com/pricing/#self-managed) which can be as high as $99 per user per month when billed annually (so almost $1.2k annually per user). Is there some sort of deal that it will be free for OpenWrt? If not, my concern that package maintainers like me will be left without merge access to luci/packages if it's going to cost substantial money to grant it.
3. Just out of curiosity, who will actually own the instance (as in be the billing/admin contact for the service provider where it's hosted and gitlab)? Would it be a private individual or an org?

PS. I'm sure you're aware, but AWS is giving free credits to FOSS developers, just throwing this out there in case someone needs some compute power for whatever reason.

**Comment by rotanid** [comment]
>     1. Is the goal to explore the possibility of moving from proprietary SaaS GitHub to own instance of gitlab?

it has already been decided to do that, so no exploration anymore.


>     2. AFAIK even self-hosted gitlab instances require [monthly per-user fees](https://about.gitlab.com/pricing/#self-managed) which can be as high as $99 per user per month when billed annually (so almost $1.2k annually per user). Is there some sort of deal that it will be free for OpenWrt? If not, my concern that package maintainers like myself will be left without merge access to luci/packages if it's going to cost substantial money to grant it.

it's always about the Open Sorce self-hosted version, not a paid version. The link you provided also shows this as the "Core" variant. Therefore, your concerns don't apply.


>     3. Just out of curiosity, who will actually own the instance (as in be the billing/admin contact for the service provider where it's hosted and gitlab)? Would it be a private individual or an org?

no billing for GitLab, and infrastructure like the other servers - but the latter is only my assumption.

note: i'm not involved, i'm just reading the mailing lists and get my info from there over the months/years.

**Comment by ynezz** [comment]
@stangri last GitLab status update http://lists.infradead.org/pipermail/openwrt-adm/2020-February/001322.html

> AWS is giving free credits to FOSS developers

Nice, where could one (or project?) apply? Quick googling didn't revealed anything usable.

**Comment by stangri** [comment]
> > AWS is giving free credits to FOSS developers
> 
> Nice, where could one (or project?) apply? Quick googling didn't revealed anything usable.

@ynezz https://aws.amazon.com/blogs/opensource/aws-promotional-credits-open-source-projects/ scroll down to how do I apply. I don't think they explicitly state how much credit they give out, but from my experience, it's generous (I'm using a fraction of what I've been given). And if you need something very specific and very high-price, I would indicate that on the application form.

If applying for a project, it may be worth your while to have the project registered as a non-profit. I'm helping out some local non-profits with their IT needs and between Microsoft/Azure and Google you can get a lot of free credits/services for cloud computing, email/office, marketing, etc. At least for USA/Canada non-profits, programs have their own qualifications for international non-profits.

If OpenWrt in general can benefit from some of that stuff above I'll be happy to look into it.

**Comment by claell** [comment]
@stangri GitLab has free offers for open source projects as well: https://about.gitlab.com/solutions/open-source/program/ so the costs you mentioned wouldn't even apply at higher GitLab tiers. That would even include hosting etc. however I noticed that most open source projects prefer to pay for their servers and host an own version theirselves.

@ynezz Thanks for linking to that mailing list entry. Personally I am not really into mailing lists, so I would have appreciated some discussion in this issue instead. I thought that this would be the "single point of truth" instead of having information spread across several platforms. Reading that it has been voted to keep the current mailing list "workflow" makes me feel a bit sad. It makes it hard to keep track of discussions etc., I'd prefer to discuss everything in issues or something like that, so topics are much easier to track for outsiders while there is still the option to reply and follow all issues via email.

**Comment by tapper82** [comment]
Gitlab is not good with screen readers.

EG on the dashboard for OpenWrt I get this lot what ever this crap is trying to tell me.
//Snip//
 <span class="repository-language-bar-tooltip-language">C</span>&nbsp;<span class="repository-language-bar-tooltip-share">66.7%</span>
<span class="repository-language-bar-tooltip-language">Makefile</span>&nbsp;<span class="repository-language-bar-tooltip-share">13.8%</span>
<span class="repository-language-bar-tooltip-language">Shell</span>&nbsp;<span class="repository-language-bar-tooltip-share">6.8%</span>
<span class="repository-language-bar-tooltip-language">Roff</span>&nbsp;<span class="repository-language-bar-tooltip-share">4.6%</span>
<span class="repository-language-bar-tooltip-language">C++</span>&nbsp;<span class="repository-language-bar-tooltip-share">3.8%</span>
/Snip/
Having to listen to my synth read that out all the time will get old quickly!
What's rong with git hub anyway?

**Comment by tapper82** [comment]
Never mind I just read ynezz commented on 18 Jul 2019

**Comment by ynezz** [comment]
> Gitlab is not good with screen readers.

@tapper82 just report it https://gitlab.com/gitlab-org/gitlab/-/issues

**Comment by adrianschmutzler** [comment]
> EG on the dashboard for OpenWrt I get this lot what ever this crap is trying to tell me.
> //Snip//
> C 66.7%
> Makefile 13.8%
> Shell 6.8%
> Roff 4.6%
> C++ 3.8%
> /Snip/

FYI, that's an overview over the different files in the project/repo, and to which kind of programming language they are assigned to by an algorithm of the platform.

This allows to classify projects for the platform, e.g. a project can be immediately categorized as "mainly C", "mainly PHP", "mainly shell" etc.

**Comment by aparcar** [comment]
Because parts of the mail archive are missing

Subject: GitLab preparation [Was: Re: Voting results]
Petr Štetiar <ynezz@true.cz> [2020-02-09 17:32:46]:
http://sprunge.us/6rfmh0

**Comment by aparcar** [comment]
We decided against GitLab and to evaluate alternatives or stick with GitHub, let's close this.
